main/lsp/requests/rename.cc

```diff
@@ -111,67 +114,73 @@ unique_ptr<WorkspaceEdit> Renamer::buildEdit() {
             newsrc = replaceMethodNameInDef(source);
-        } else { // send
-            newsrc = replaceMethodNameInSend(source);
+        } else {
+            ENFORCE(0, "Unexpected query response type while renaming method");
+            return;
         }
         edits[location->uri].push_back(make_unique<TextEdit>(move(location->range), newsrc));
     }
 
 private:
-    // Parse a method name from a send (call), replace it with the new name, and return the full send source with the
-    // new name
-    string replaceMethodNameInSend(string send) {
-        // Split into method and args
-        string methodExpr, args;
-        for (string::size_type i = 0; i < send.size(); i++) {
-            if (send[i] == ' ' || send[i] == '(') {
-                methodExpr = send.substr(0, i);
-                args = send.substr(i, send.size() - i);
+    string replaceMethodNameInDef(string def) {
+        // We can't do a simple string replace-all because of cases like:
+        //   `def foo(foobar)`
+        //   `attr_reader :att`
+        // Since we don't have more precise info about where the name is, we do this:
+        //   1. if the definition starts with attr_{reader,writer,accessor} then skip over that
+        //   2. and then change the first instance of oldName to newName
+        string::size_type offset = 0;
+        vector<string> prefixes{"attr_reader", "attr_accessor", "attr_writer"};
+        for (auto prefix : prefixes) {
+            if (absl::StartsWith(def, prefix)) {
+                offset = prefix.length();
                 break;
             }
         }
-        // no args passed
-        if (methodExpr.size() == 0) {
-            methodExpr = send;
-            args = "";
-        }
-        ENFORCE(send == (methodExpr + args));
-
-        // The method name must be at the end of the methodExpr; if so, replace it and reconstruct the send
-        auto prefixLen = methodExpr.size() - oldName.size();
-        if (oldName == methodExpr.substr(prefixLen, oldName.size())) {
-            auto newMethodExpr = methodExpr.substr(0, prefixLen) + newName;
-            return (newMethodExpr + args);
-        } else {
-            // If this happens we parsed the send incorrectly (or we parsed something that's not a send), so give up and
-            // return the unmodified source
-            ENFORCE(0);
-            return send;
+        auto pos = def.find(oldName, offset);
+        if (pos == string::npos) {
+            ENFORCE(0, "Method name should appear in definition");
+            return def;
         }
+        return replaceAt(def, pos);
     }
-
-    string replaceMethodNameInDef(string def) {
-        return absl::StrReplaceAll(def, {{oldName, newName}});
+    string replaceAt(string input, string::size_type pos) {
+        auto suffixOffset = pos + oldName.length();
+        return input.substr(0, pos) + newName + input.substr(suffixOffset, input.length() - suffixOffset);
```

Comment:
Super micro optimization: This will reduce the number of intermediate string allocations (concat-ing string views).

```suggestion
    string replaceAt(string_view input, string::size_type pos) {
        auto suffixOffset = pos + oldName.length();
        return absl::StrCat(input.substr(0, pos), newName, input.substr(suffixOffset, input.length() - suffixOffset));
```